src/sockaddr.rs

```diff
@@ -27,153 +16,86 @@ pub struct SockAddr {
             len: len,
         }
     }
 
-    /// Constructs a `SockAddr` with the family `AF_UNIX` and the provided path.
-    ///
-    /// This function is only available on Unix.
-    ///
-    /// # Failure
-    ///
-    /// Returns an error if the path is longer than `SUN_LEN`.
-    #[cfg(all(unix, feature = "all"))]
-    pub fn unix<P>(path: P) -> ::std::io::Result<SockAddr>
-    where
-        P: AsRef<::std::path::Path>,
-    {
-        use libc::{c_char, sockaddr_un, AF_UNIX};
-        use std::cmp::Ordering;
-        use std::io;
-        use std::os::unix::ffi::OsStrExt;
-
-        unsafe {
-            let mut addr = mem::zeroed::<sockaddr_un>();
-            addr.sun_family = AF_UNIX as sa_family_t;
-
-            let bytes = path.as_ref().as_os_str().as_bytes();
-
-            match (bytes.get(0), bytes.len().cmp(&addr.sun_path.len())) {
-                // Abstract paths don't need a null terminator
-                (Some(&0), Ordering::Greater) => {
-                    return Err(io::Error::new(
-                        io::ErrorKind::InvalidInput,
-                        "path must be no longer than SUN_LEN",
-                    ));
-                }
-                (Some(&0), _) => {}
-                (_, Ordering::Greater) | (_, Ordering::Equal) => {
-                    return Err(io::Error::new(
-                        io::ErrorKind::InvalidInput,
-                        "path must be shorter than SUN_LEN",
-                    ));
-                }
-                _ => {}
-            }
-
-            for (dst, src) in addr.sun_path.iter_mut().zip(bytes) {
-                *dst = *src as c_char;
-            }
-            // null byte for pathname is already there since we zeroed up front
+    /// Returns this address's family.
+    pub fn family(&self) -> sa_family_t {
+        self.storage.ss_family
+    }
 
-            let base = &addr as *const _ as usize;
-            let path = &addr.sun_path as *const _ as usize;
-            let sun_path_offset = path - base;
+    /// Returns the size of this address in bytes.
+    pub fn len(&self) -> socklen_t {
+        self.len
+    }
 
-            let mut len = sun_path_offset + bytes.len();
-            match bytes.get(0) {
-                Some(&0) | None => {}
-                Some(_) => len += 1,
-            }
-            Ok(SockAddr::from_raw_parts(
-                &addr as *const _ as *const _,
-                len as socklen_t,
-            ))
-        }
+    /// Returns a raw pointer to the address.
+    pub fn as_ptr(&self) -> *const sockaddr {
+        &self.storage as *const _ as *const _
     }
 
-    unsafe fn as_<T>(&self, family: sa_family_t) -> Option<T> {
-        if self.storage.ss_family != family {
-            return None;
+    /// Returns this address as a `SocketAddr` if it is in the `AF_INET` (IP v4)
+    /// or `AF_INET6` (IP v6) family, otherwise returns `None`.
+    pub fn as_std(&self) -> Option<SocketAddr> {
```

Comment:
This can be a safe one-liner :)

```rust
self.as_inet().or_else(|| self.as_inet6())
```

Reply:
We need to call `Into::into` first, so we'll get.

```rust
        self.as_inet()
            .map(|a| a.into())
            .or_else(|| self.as_inet6().map(|a| a.into()))
```

Not quite a one-liner, but if you prefer it I'll change it.

Reply:
Up to you - it reduces unsafe code a bit though.